TwoPaneLayout/library/src/main/java/com/microsoft/device/display/twopanelayout/TwoPaneLayoutImpl.kt

```diff
@@ -0,0 +1,286 @@
+                maxWidth = (maxWidth * ratio).toInt(),
+                maxHeight = maxHeight
+            )
+        } else {
+            Constraints(
+                minWidth = minWidth,
+                minHeight = (minHeight * ratio).toInt(),
+                maxWidth = maxWidth,
+                maxHeight = (maxHeight * ratio).toInt()
+            )
+        }
+
+        val placeable = measurables[i].measure(childConstraints)
+        placeables.add(placeable)
+    }
+    return placeables
+}
+
+private fun Placeable.PlacementScope.placeTwoPaneEqually(
+    orientation: LayoutOrientation,
+    placeable: Placeable,
+    index: Int,
+    paneWidth: Int,
+    paneHeight: Int,
+    paddingBounds: Rect,
+    constraints: Constraints
+) {
+    if (orientation == LayoutOrientation.Vertical) {
+        var xPosition = 0 // for the first pane
+        if (index != 0) { // for the second pane
+            var lastPaneWidth = paneWidth - paddingBounds.right
+            var firstPaneWidth = constraints.maxWidth - lastPaneWidth
+            xPosition += firstPaneWidth
+        }
+        placeable.placeRelative(x = xPosition, y = 0)
+    } else {
+        var yPosition = 0
+        if (index != 0) {
+            var lastPaneHeight = paneHeight - paddingBounds.bottom
+            var firstPaneHeight = constraints.maxHeight - lastPaneHeight
+            yPosition += firstPaneHeight
+        }
+        placeable.placeRelative(x = 0, y = yPosition)
+    }
+}
+
+private fun Placeable.PlacementScope.placeTwoPaneProportionally(
+    orientation: LayoutOrientation,
+    placeable: Placeable,
+    index: Int,
+    twoPaneParentData: Array<TwoPaneParentData?>,
+    constraints: Constraints,
+    totalWeight: Float
+) {
+    if (orientation == LayoutOrientation.Vertical) {
+        var xPosition = 0 // for the first pane
+        if (index != 0) { // for the second pane
+            val parentData = twoPaneParentData[index]
+            val weight = parentData.weight
+            val ratio = 1f - (weight / totalWeight)
+            var firstPaneWidth = (constraints.maxWidth * ratio).toInt()
+            xPosition += firstPaneWidth
+        }
+        placeable.placeRelative(x = xPosition, y = 0)
+    } else {
+        var yPosition = 0
+        if (index != 0) {
+            val parentData = twoPaneParentData[index]
+            val weight = parentData.weight
+            val ratio = 1f - (weight / totalWeight)
+            var firstPaneHeight = (constraints.maxHeight * ratio).toInt()
+            yPosition += firstPaneHeight
+        }
+        placeable.placeRelative(x = 0, y = yPosition)
+    }
+}
+
+private val IntrinsicMeasurable.data: TwoPaneParentData?
+    get() = parentData as? TwoPaneParentData
+
+private val TwoPaneParentData?.weight: Float
+    get() = this?.weight ?: 0f
+
+/**
+ * Parent data associated with children.
+ */
+internal data class TwoPaneParentData(
+    var weight: Float = 0f,
+)
+
+internal class LayoutWeightImpl(
```

Comment:
should we use here a `data class`? since you are overriding later `equals`, `hashcode`, etc. maybe is worth?